ouroboros-network/src/Ouroboros/Network/PeerSelection/EstablishedPeers.hs

```diff
@@ -0,0 +1,231 @@
+         establishedStatus = Map.delete peeraddr establishedStatus,
+         nextActivateTimes = PSQ.delete peeraddr nextActivateTimes }
+
+
+
+-- | Bulk delete of peers from 'EstablishedPeers.
+--
+deletePeers :: Ord peeraddr
+            => Set peeraddr
+            -> EstablishedPeers peeraddr peerconn
+            -> EstablishedPeers peeraddr peerconn
+deletePeers peeraddrs es@EstablishedPeers { establishedReady,
+                                            establishedStatus,
+                                            nextActivateTimes } =
+    es { establishedReady  = foldl' (flip Map.delete) establishedReady  peeraddrs,
+         establishedStatus = foldl' (flip Map.delete) establishedStatus peeraddrs,
+         nextActivateTimes = foldl' (flip PSQ.delete) nextActivateTimes peeraddrs }
+
+
+--
+-- Time managment
+--
+
+setCurrentTime :: Ord peeraddr
+               => Time
+               -> EstablishedPeers peeraddr peerconn
+               -> EstablishedPeers peeraddr peerconn
+setCurrentTime now ep@EstablishedPeers { establishedReady,
+                                         nextActivateTimes } =
+    let ep' = ep { establishedReady =
+                        Map.fromList [ (peeraddr, peerconn) 
+                                     | (peeraddr, _, peerconn) <- nowAvailableToActivate ]
+                     <> establishedReady,
+                   nextActivateTimes = nextActivateTimes' }
+    in assert (invariant ep') ep'
+  where
+    (nowAvailableToActivate, nextActivateTimes') =
+      PSQ.atMostView now nextActivateTimes
+
+
+minActivateTime :: Ord peeraddr
+                => EstablishedPeers peeraddr peerconn
+                -> Maybe Time
+minActivateTime EstablishedPeers {
+                   establishedReady,
+                   nextActivateTimes
+                 }
+  | Map.null establishedReady
+  , Just (_k, t, _, _psq) <- PSQ.minView nextActivateTimes
+  = Just t
+
+  | otherwise
+  = Nothing
+
+
+setActivateTime :: Ord peeraddr
+                => Set peeraddr
+                -> Time
+                -> EstablishedPeers peeraddr peerconn
+                -> EstablishedPeers peeraddr peerconn
+setActivateTime peeraddrs _time ep | Set.null peeraddrs = ep
+setActivateTime peeraddrs time  ep =
+    let ep' = foldl'
+                (\peers@EstablishedPeers { establishedReady  = ready,
+                                           nextActivateTimes = times }
+                  peeraddr ->
+
+                  -- 'alterF' allows to delete and return the value
+                  case Map.alterF (\x -> (x, Nothing)) peeraddr ready of
+                    (Just peerconn, ready') ->
+                      peers { establishedReady  = ready',
+                              nextActivateTimes = PSQ.insert peeraddr time peerconn times }
+
+                    (Nothing, _ready') ->
+                      error "EstablishedPeers.setActivateTime: invariant violation")
+                ep
+                peeraddrs
```

Comment:
Unfortunately, we cannot. Here, unlike in `setConnectTime` or `setGossipTime` we push the `peerconn` that was removed from `establishedReady` to `nextActivateTimes`.  There's no such depenency in the other two (they just put `()` as the value in the priority queue).

Reply:
After changing the internal representation to store all connections instead of just the ready ones in the map (as discussed below), I also simplified this function. Now we no longer need to drag `peerconn` around.